### PR TITLE
Accessing and modifying description and labels of Enso Cloud assets

### LINKED ISSUE #11227: Access metadata of Enso Cloud assets programmatically
We want to be able to check the description and labels of an Enso Cloud asset through Enso code.

To do so we want to extend the `Enso_File` API:
- [x] add `description : Text` method that inspects description
- [x] add `labels : Vector ?` method that returns the list of labels associated with the asset.
	- [x] figure out how the labels should be represented: just `Text` or a custom type?
	- [x] do we need special handling to see if a label with given name exists and if not, create it?

Additionally we want to be able to set the description and labels:
- [x] add `set_description`
- [x] add `set_labels` that replaces the set of labels
- [x] add `add_label` that just adds a new label, keeping any other ones

It was also suggested that we want an easy way for users to get the metadata with `Data.read`. We may consider adding a special 'file format' `Enso_File_Metadata` such that if I read `Data.read "enso://~/my-file.txt" ..Enso_File_Metadata` then I will get a special object (or just JSON) that contains the description and the labels.

### LINKED ISSUE #11278: Often random failures of `Audit_Spec` on the CI
Numerous reports about spontaneous failures of Audit_Spec on CI have been reported:
```
 INFO ide_ci::program::command: bash ℹ️ [FAILED] [PostgreSQL] Audit Logs: [2/3, 112884ms]
 INFO ide_ci::program::command: bash ℹ️ 
 INFO ide_ci::program::command: bash ℹ️     - [FAILED] should see Database operations performed on a table through the standard workflow [112422ms]
 INFO ide_ci::program::command: bash ℹ️         Reason: An unexpected dataflow error (Not_Found) has been matched (at /runner/_work/enso/enso/test/Table_Tests/src/Database/Common/Audit_Spec.enso:44:17-37).
 INFO ide_ci::program::command: bash ℹ️         at <enso> <default::Vector.find::if_missing>(/runner/_work/enso/enso/built-distribution/enso-engine-2024.5.1-dev-linux-amd64/enso-2024.5.1-dev/lib/Standard/Base/2024.5.1-dev/src/Data/Vector.enso:407:100-120)
```

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

- Closes #11227
- Additionally, it should fix #11278 by ensuring that every scheduled message goes to the desired endpoint, by splitting each batch by endpoint.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso [hunk 1]
```diff
@@ -138,12 +139,123 @@ type Enso_File
...
+       ? Race Conditions
+
+         Removing labels is not atomic. If two processes are modifying labels of
+         the same asset at the same time, some changes may be lost.
+    remove_label : Text -> Boolean
+    remove_label self (label : Text) -> Boolean =
+        Context.Output.if_enabled disabled_message="Cannot remove a label when writing is disabled. Press the Write button ▶ to perform the operation." panic=False <|
+            labels = self.labels
+            if labels.contains label . not then False else
+                new_labels = labels.filter l-> l != label
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso:225]: It would be nice to have this as `Vector.remove_element`.

**Review by GregoryTravis — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso [hunk 2]
```diff
@@ -138,12 +139,123 @@ type Enso_File
...
+         Adding labels is not atomic. If two processes are modifying labels of
+         the same asset at the same time, some changes may be lost.
+    add_label : Text -> Enso_File
+    add_label self (label : Text) -> Enso_File =
+        Context.Output.if_enabled disabled_message="Cannot add a label when writing is disabled. Press the Write button ▶ to perform the operation." panic=False <|
+            labels = self.labels
+            if labels.contains label then self else
+                ensure_tags_exist [label] . if_not_error <|
+                    new_labels = labels+[label]
+                    update_asset_labels self new_labels . if_not_error self
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso]:
```suggestion
        labels = self.labels
        if labels.contains label then self else
            ensure_tags_exist [label] . if_not_error <|
                new_labels = labels+[label]
                Context.Output.if_enabled disabled_message="Cannot add a label when writing is disabled. Press the Write button ▶ to perform the operation." panic=False <|
                    update_asset_labels self new_labels . if_not_error self
```

Could make it only check the context as last step then allow user to fix issue

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso (hunk 1 above)
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso]:
```suggestion
        labels = self.labels
        if labels.contains label . not then False else
            Context.Output.if_enabled disabled_message="Cannot remove a label when writing is disabled. Press the Write button ▶ to perform the operation." panic=False <|
                        new_labels = labels.filter l-> l != label
```

as above

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso [hunk 3]
```diff
@@ -138,12 +139,123 @@ type Enso_File
...
+
+       Returns:
+       - The file on which it was called.
+
+       ? Race Conditions
+
+         Adding labels is not atomic. If two processes are modifying labels of
+         the same asset at the same time, some changes may be lost.
+    add_label : Text -> Enso_File
+    add_label self (label : Text) -> Enso_File =
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso:198]: worth routing this through `set_labels`?

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso [hunk 4]
```diff
@@ -138,12 +139,123 @@ type Enso_File
...
+       Arguments:
+       - labels: The new set of labels to associate with the file.
+
+       Returns:
+       - The file on which it was called.
+    set_labels : Vector Text -> Enso_File
+    set_labels self (labels : Vector Text) -> Enso_File =
+        Context.Output.if_enabled disabled_message="Cannot update labels when writing is disabled. Press the Write button ▶ to perform the operation." panic=False <|
+            ensure_tags_exist labels . if_not_error <|
+                update_asset_labels self labels . if_not_error self
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso]: as above could do context check late

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso [hunk 5]
```diff
@@ -138,12 +139,123 @@ type Enso_File
...
+       ICON data_output
+       Creates a new label with the given name and color.
+
+       Arguments:
+       - name: The name of the label to create.
+       - color: The color of the label to create.
+    create_label : Text -> Text -> Nothing
+    create_label (name : Text) (color : Color) =
+        Context.Output.if_enabled disabled_message="Cannot create a label when writing is disabled. Press the Write button ▶ to perform the operation." panic=False <|
+            create_tag name color . if_not_error Nothing
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso]: worth testing existance?

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Existing_Enso_Asset.enso [hunk 6]
```diff
@@ -100,36 +113,71 @@ type Existing_Enso_Asset
     from_json json -> Existing_Enso_Asset =
         title = get_required_field "title" json expected_type=Text
         id = get_required_field "id" json expected_type=Text
-        Existing_Enso_Asset.from_id_and_title id title
+        metadata = Asset_Metadata.from_json json
+        Existing_Enso_Asset.new_with_metadata id title metadata
+
+    ## PRIVATE
+    new_with_metadata id:Text title:Text metadata:Asset_Metadata -> Existing_Enso_Asset =
+        Existing_Enso_Asset.Value title id (asset_type_from_id id) metadata
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Existing_Enso_Asset.enso]:
```suggestion
    ## PRIVATE
    new id:Text title:Text metadata:(Nothing|Asset_Metadata)=Nothing -> Existing_Enso_Asset =
        Existing_Enso_Asset.Value title id (asset_type_from_id id) metadata
```

Could we not have a single `new` here?

**Review by jdunkerley — APPROVED** [review]
Generally looks good - a few suggestions.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso (hunk 2 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso]: No, we shouldn't change the order. Perhaps I should rename the `ensure_tags_exist` method to make it clearer.

This method **creates** tags if they don't exist. So it is a method with side-effects, that calls into the Cloud. Hence, it itself should be behind the Output context.

I will rename the method to `create_tags_if_not_exist` to make it clearer that this is more than a pre-check.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso (hunk 5 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso]: OK. Although ideally that should be checked on the cloud side and only a special error returned.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Existing_Enso_Asset.enso (hunk 6 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Existing_Enso_Asset.enso]: I wanted to make it very clear when the asset is constructed with or without metadata. IMO it is a bit clearer when this is verbose. But I can merge them - initially there were more differences in the logic, now indeed it's very similar

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso (hunk 3 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso:198]: I was thinking of it, but then the message if Context was disabled would be 'Cannot set labels...', but I wanted a more appropriate/precise 'Cannot add a label'.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso (hunk 2 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso]: But I can indeed move the `contains` check before context check 👍 

Sorry at first I did not notice this part

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso (hunk 1 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso:225]: Indeed, but adding it I'd have to add unit tests for it and I don't want to postpone this PR.

Let's add it in a separate PR when we see it used more, or see a need for it for end-users.

**Pull request merged**